KEEP faucet: I grant thee

sthompson22 opened the pull request:
### Intro

Here we provide issuing KEEP grants expressed as a JS Google Cloud Function.  The purpose here is to provide a minimally functional faucet for issuing a KEEP grant, while keeping operational overhead in mind. Minimally functional here being that for a given account we can issue
a KEEP grant on demand from a public interface.

### Why Google Functions?

I took stock of my motivations around implementation, selfishly these are pretty heavily tilted towards administration of the faucet....but I think we landed in an ok place.

My main motivators were:

- Observable
- Scalable
- Controllable
- Minimal operational overhead

#### Observable

With GCF (Google Cloud Functions) we get some decent observability out of the box.  Error logs and tracking, invocation counts, etc.  tl;dr here is that we can get a snapshot of the faucet state viewing a single page on the Google Cloud Console.  We can also setup alerts around logs / error conditions.

#### Scalable

We're in infinite scalability land here.  GCF is great in that each request to the faucet spawns its own instance of funding/staking a given address.

#### Controllable

We don't want people draining all of our KEEP,  DoS'ing the faucet, or costing us a ton of money in infra.  GCF's are are nice here in that we can (as part of function configuration), limit the number of in-flight function invocations.  This helps control cost and keeps a person from spinning up N > 5 requests at once.

What's missing here is throttling on the faucet endpoint.  For that we need to integrate this cloud function with Google Cloud Endpoints.  My plan was to punt on that for V1, but I'm open for discussion.

#### Minimal Operational Overhead

- We can do the thing in a few lines of JS.
- No infra to run or maintain.  
- Local development is facilitated by a single NPM package (example below).
- Build and release flow can be facilitated totally by Circle.
- GCF gives you versioning out of the box.

#### Note on cost

This should be near free for us to operate.  Here's the free tier stats for Google Cloud Functions:

> Cloud Functions provides a perpetual free tier for compute-time resources, which includes an allocation of both GB-seconds and GHz-seconds. In addition to the 2 million invocations, the free tier provides 400,000 GB-seconds, 200,000 GHz-seconds of compute time and 5GB of Internet egress traffic per month. Note that even for free tier usage, we require a valid billing account.

### Grant Amounts

Base grant value is 300k KEEP.  The function will check if you're at that balance or higher, if you are it will exit.  Otherwise it will ship a 300k grant.

### Local Testing

GCF's have a pretty sweet [testing framework](https://cloud.google.com/functions/docs/functions-framework) for iterating locally.  You can see that reflected in the `package.json` checked in here, specifically:

```
"devDependencies": {
    "@google-cloud/functions-framework": "^1.3.2"
  },
  "scripts": {
    "start": "functions-framework --target=issueGrant"
  }
```

You can then test your function by running `npm start` on the root dir of the GCF.  When you do it will look something like this:  (don't forget `npm i` first)

```
sthompson22 ~/Projects/keep/sandbox/google-functions/keep-faucet $ npm start

> @ start /Users/sthompson22/Projects/keep/sandbox/google-functions/keep-faucet
> functions-framework --target=issueGrant


Serving function...
Function: issueGrant
URL: http://localhost:8080/
```

From here you can invoke your function by curling `localhost:8080`.  Here's an example of me testing this function:

#### Account not at max allowed

```
sthompson22 ~/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet (sthompson22/keep-test/KEEP-faucet) $ curl localhost:8080/'?account=0xcCb2acbC6db2432d577b9EdBD42c1B125ef9925e'

        Created token grant with 1000000000000000000 KEEP for account: 0xcCb2acbC6db2432d577b9EdBD42c1B125ef9925e
        You can manage your token grants at: https://dashboard.test.keep.network
        You can find us on Discord at: https://discord.gg/jqxBU4m
```

#### Account at max

```
sthompson22 ~/Projects/keep/keep-core/infrastructure/gcp/keep-test/google-functions/keep-faucet (sthompson22/keep-test/KEEP-faucet) $ curl localhost:8080/'?account=0xfeC11102F4b141975FF0e9471203F36e1ce17E4E'

        Token grant failed, your account has the maximum testnet KEEP allowed.
        You can manage your token grants at: https://dashboard.test.keep.network
        If you have questions find us on Discord: https://discord.gg/jqxBU4m
```

### Function Deployment (To Google Cloud Platform)

For now I've been deploying the function from local via `gcloud`.  This can be integrated with Circle and should be.  Not sure I'll do that on this pass, open to discussion.

Example deploy command:

```
gcloud functions deploy keep-faucet-test --trigger-http --runtime nodejs10 --entry-point dripAndStake --set-env-vars KEEP_CONTRACT_OWNER_ADDRESS=0x0396457C928e58axxx,KEEP_CONTRACT_OWNER_PRIVATE_KEY=EBAE221Dxxxx,ETHEREUM_HOST=https://ropsten.infura.io/v3/59fb36a36fa4474b890c1xxxx,ETHEREUM_NETWORK_ID=3
```

### Calling the Faucet

Once the function is deployed it's given a public endpoint.  You can hit that endpoint via curl or a browser.  To feed the Ethereum account that is to be issued a KEEP grant you need to pass `?account=<my-account-address>` with the function URL.  e.g. To test out the deployed test faucet you can hit:





### Linting / Prettier 

I've added a global `pre-commit-hook` for `prettier`.  After running `prettier` I ran `js:lint` and `js:lint:fix` to make sure formatting between the two weren't at odds.

sthompson22 commented:
This is about to get a facelift.

sthompson22 commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/issue-grant.js:
```diff
@@ -0,0 +1,117 @@
+        )
+        .send({ from: keepContractOwnerAddress })
+
+      console.log(`Created grant for ${web3.utils.toBN(grantAmount)} to: ${granteeAccount}`)
```
This doesn't display quite right as we still have the 18 `0` offset.

sthompson22 replied:
Still true.

Shadowfiend commented on .pre-commit-config.yaml:
```diff
@@ -1,28 +1,28 @@
+   rev: "master" # Use the sha or tag you want to point at
+   hooks:
+    - id: prettier
+      files: '\.js$'
```
Lol what happened here.

Shadowfiend replied:
Ah. No, we don't want prettier here, we want eslint… Here's how we do this in keep-ecdsa: https://github.com/keep-network/keep-ecdsa/blob/master/.pre-commit-config.yaml#L14-L19 .

Our eslint config also handles prettier formatting.

Shadowfiend commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/parse-account-address.js:
```diff
@@ -3,18 +3,18 @@ const url = require('url')
-  } else if (!/^(0x)?[0-9a-f]{40}$/i.test(address)) {
-    // check if it has the basic requirements of an address
+  if (requestUrl.query.account) {
+    return (account = requestUrl.query.account)
```
`account` is never declared anywhere, and we never use it, right? I think you can just directly return here.

Shadowfiend commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/parse-account-address.js:
```diff
@@ -3,18 +3,18 @@ const url = require('url')
 
 exports.parseAccountAddress = (request, response) => {
   requestUrl = url.parse(request.url, true)
```
This is never declared. `const requestUrl = …`.

sthompson22 replied:
oh my.

Shadowfiend commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/parse-account-address.js:
```diff
@@ -3,18 +3,18 @@ const url = require('url')
-    // check if it has the basic requirements of an address
+  if (requestUrl.query.account) {
+    return (account = requestUrl.query.account)
+  } else if (!/^(0x)?[0-9a-f]{40}$/i.test(account)) {
```
`account` is never defined here. Mayhaps after we declare `requestUrl` we can `const account = requestUrl.query.account` and then use `account` throughout the rest of the function.

sthompson22 replied:
oh my^2

Shadowfiend commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/parse-account-address.js:
```diff
@@ -3,18 +3,18 @@ const url = require('url')
     // double thanks to the Ethereum folks for figuring this regex out already
     return response.send(
-      'Improperly formatted account address, please try a valid one.',
+      'Improperly formatted account account, please try a valid one.',
```
Probably `account address` is still correct here.

Shadowfiend commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/parse-account-address.js:
```diff
@@ -3,18 +3,18 @@ const url = require('url')
+    console.log('No account set in HTTP request.')
     return response.send(
-      'No account address set, please set an address with ?address=<accountAddress>',
+      'No account account set, please set an account with ?account=<accountAddress>',
```
Ditto on `account address`, perhaps.

sthompson22 commented:
Thanks, addressed the comments.  We bubble up properly now but `issueGrant` tries to continue despite hitting. the error case in `parse-account-address`

Shadowfiend commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/package.json:
```diff
@@ -0,0 +1,19 @@
+  "dependencies": {
+    "@keep-network/keep-core": ">0.12.0-rc <0.12.0",
+    "@truffle/hdwallet-provider": "^1.0.25",
+    "web3": "1.0.0-beta.55"
```
This can be `1.2.6`.

Shadowfiend commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/parse-account-address.js:
```diff
@@ -0,0 +1,21 @@
+exports.parseAccountAddress = (request, response) => {
+  const requestUrl = url.parse(request.url, true)
+  const account = requestUrl.query.account
+  if (requestUrl.query.account) {
```
This can just be `if (account)` now.

sthompson22 replied:
derp

sthompson22 commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/package.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "main": "issue-grant.js",
+  "dependencies": {
+    "@keep-network/keep-core": ">0.12.0-rc <0.12.0",
```
Yep - thought about just going and and roping this into CI but this thing is getting real long in the tooth.

sthompson22 commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/.eslintrc.js:
```diff
@@ -0,0 +1,12 @@
+	],
+	'globals': {
+		'Atomics': 'readonly',
+		'SharedArrayBuffer': 'readonly'
```
This file is gone now.

sthompson22 commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/.eslintrc.js:
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+	"parser": "babel-eslint",
```
File ded.

sthompson22 commented on .prettierrc.yaml:
```diff
@@ -0,0 +1,3 @@
+semi: false
+trailingComma: "all"
+singleQuote: true
```
File dedddd.

sthompson22 commented on infrastructure/gcp/keep-test/google-functions/keep-faucet/issue-grant.js:
```diff
@@ -0,0 +1,110 @@
+      response.send(`
+        Created token grant with ${web3.utils.toBN(
+          grantAmount,
+        )} KEEP for account: ${granteeAccount}
```
No not really.

```
        Created token grant with 300000000000000000000000 KEEP for account: 0x0F0977c4161a371B5E5eE6a8F43Eb798cD1Ae1DB
        You can manage your token grants at: https://dashboard.test.keep.network
        You can find us on Discord at: https://discord.gg/jqxBU4m
```

Sample ouput.

sthompson22 commented:
Thanks @Shadowfiend pushed a few updates per your comments - namely got rid of the bastard prettier config here in favor of what we do in `keep-ecdsa`.

Shadowfiend commented on .pre-commit-config.yaml:
```diff
@@ -21,8 +21,9 @@ repos:
+      entry: /usr/bin/env bash -c "cd solidity && npm run lint:js"
       files: '\.js$'
+      language: script
+      description: "Checks JS code according to the package's linter configuration"
```
I guess we probably want to add another one here to run on your functions. Wonder how we could generalize that… Anyway, for another time.
